src/libraries/System.Runtime.Serialization.BinaryFormat/src/PayloadReader.cs

```diff
@@ -0,0 +1,285 @@
+{
+    private static UTF8Encoding ThrowOnInvalidUtf8Encoding { get; } = new(false, throwOnInvalidBytes: true);
+
+    /// <summary>
+    /// Checks if given buffer contains only Binary Formatter payload.
+    /// </summary>
+    /// <param name="bytes">The buffer to inspect</param>
+    /// <returns>True if the first and last bytes indicate Binary Format, otherwise false.</returns>
+    public static bool ContainsBinaryFormatterPayload(ReadOnlySpan<byte> bytes)
+        => bytes.Length >= SerializedStreamHeaderRecord.Size + 2
+            && bytes[0] == (byte)RecordType.SerializedStreamHeader
+            && bytes[bytes.Length - 1] == (byte)RecordType.MessageEnd;
+
+    /// <summary>
+    /// Checks if given stream contains only Binary Formatter payload.
+    /// </summary>
+    /// <param name="stream">The readable and seekable stream to inspect.</param>
+    /// <returns>True if the first and last bytes indicate Binary Format, otherwise false.</returns>
+    /// <exception cref="ArgumentNullException">The stream is null.</exception>
+    /// <exception cref="NotSupportedException">The stream does not support reading or seeking.</exception>
+    /// <exception cref="ObjectDisposedException">The stream was closed.</exception>
+    /// <remarks><para>It does not modify the position of the stream.</para></remarks>
+    public static bool ContainsBinaryFormatterPayload(Stream stream)
+    {
+#if NETCOREAPP
+        ArgumentNullException.ThrowIfNull(stream);
+#else
+        if (stream is null) throw new ArgumentNullException(nameof(stream));
+#endif
+        long beginning = stream.Position;
+        if (stream.Length - beginning < SerializedStreamHeaderRecord.Size + 2)
+        {
+            return false;
+        }
+
+        try
+        {
+            int firstByte = stream.ReadByte();
+            if (firstByte != (byte)RecordType.SerializedStreamHeader)
+            {
+                return false;
+            }
+
+            stream.Seek(-1, SeekOrigin.End);
+            int lastByte = stream.ReadByte();
+            return lastByte == (byte)RecordType.MessageEnd;
+        }
+        finally
+        {
+            stream.Position = beginning;
+        }
+    }
+
+    /// <summary>
+    /// Reads the provided Binary Format payload.
+    /// </summary>
+    /// <param name="payload">The Binary Format payload.</param>
+    /// <param name="options">An object that describes optional <seealso cref="PayloadOptions"/> parameters to use.</param>
+    /// <param name="leaveOpen">True to leave the <paramref name="payload"/> payload open
+    /// after the reading is finished, otherwise, false.</param>
+    /// <returns>A <seealso cref="SerializationRecord"/> that represents the root object.
+    /// It can be either <seealso cref="PrimitiveTypeRecord{T}"/>,
+    /// a <seealso cref="ClassRecord"/> or an <seealso cref="ArrayRecord"/>.</returns>
+    /// <exception cref="ArgumentNullException">When <paramref name="payload"/> is null.</exception>
```

Comment:
https://github.com/dotnet/dotnet-api-docs/wiki/Exceptions says

> Write each condition as if the word "if" precedes it. Write the description in present tense, unless past or future tense is truly necessary to convey the progression of action.

OK, we always skip the past tense part for this particular reason, but the `as if the word "if" precedes it` definitely applies.

Also... `null` => `<see langword="null" />` (which turns to `Nothing` for VB).

```suggestion
    /// <exception cref="ArgumentNullException"><paramref name="payload"/> is <see langword="null" />.</exception>
```